src/Base/PyHandle.h

```diff
@@ -0,0 +1,152 @@
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef FREECAD_BASE_PYHANDLE_H
+#define FREECAD_BASE_PYHANDLE_H
+
+#include <CXX/Objects.hxx>
+#include "PyObjectBase.h"
+
+namespace Base {
+
+/** UnsafePyHandle: a convenient way to use python references as smart pointers for C++ objects.
+ * "Unsafe" stands for lack of runtime checks: type checking is limited; dereferencing the object
+ * if the C++ object was deleted will crash the application (you can check isValid() to guard
+ * against the latter).
+ *
+ * Usage example:
+ *     //create a py object, and save it into the handle
+ *     UnsafePyHandle<Base::Placement> h (new Base::PlacementPy(new Base::Placement), true);
+ *     //use the handle as if it is a pointer to C++ object
+ *     h->getPosition()
+ *     //and when the handle is destroyed, so is the py object, if it isn't referred to by something else.
+ *
+ * ## Memory management
+ *
+ * * when PyHandle is created, reference counter is incremented (except if
+ * "new_reference" constructor argument is true).
+ *
+ * * when PyHandle is destroyed, reference counter is decremented (no matter
+ * what "new_reference" value was passed to constructor). If resulting refcount
+ * is zero, the object is deleted.
+ *
+ * * Py::new_reference_to(PyHandle) increments reference count and returns PyObject*
+ *
+ * Cheatsheet:
+ *
+ * * when constructing from "new SomethingSomethingPy(new SomethingSomething)",
+ * use "new_reference". That is because SomethingSomethingPy sets reference counter
+ * to 1 in its constructor.
+ *
+ * * when constructing from getPyObject, use "new_reference"
+ *
+ * * when constructing from PyObject* you got as an argument of ...PyImp method
+ * implementation, don't use "new_reference".
+ *
+ * * when returning PyObject* from a getPyObject, use
+ * Py::new_reference_to(<PyHangle> object).
+ *
+ * * when returning PyObject* from a ...PyImp method implementation, use
+ * Py::new_reference_to(<PyHangle> object).
+ *
+ * The handle derives from PyCXX's Py::Object, so it can be conveniently used
+ * for adding to tuples and the like, when programming python API.
+ */
+ // Since it was made to become the main memory management construct of
+ // ConstraintSolver, checks for isValid() are intentionally not done to maximize
+ // performance. It is named "Unsafe" with the intention to have room for a safer
+ // "PyHandle" replica.
+template <class CppType>
+class UnsafePyHandle : public Py::Object
+{
+private:
+    CppType* cppptr;
+protected:
+    virtual void setObject(PyObject* pyob, bool new_reference = false){
+        if (typecheck(pyob))
+            cppptr = static_cast<CppType*>(static_cast<Base::PyObjectBase*>(pyob)->twinPtr());
+        else
+            throw Base::TypeError("Not a freecad object");
+        this->set(pyob, new_reference);
+    }
+
+public:
+    virtual bool typecheck(PyObject* pyob) {
```

Comment:
why is this `virtual`? I don't see any derived class that is `override` it.

...if instead this is `override` a method from `Py::Object`, then you should explicitly state that - trust me, this will catch some really gnarly runtime-issues at compile-time for you:

```cpp
virtual bool typecheck(PyObject* pyob) override{...}
```

Reply:
this isn't an override. It is virtual to make it overridable, which is unlikely.
Py::Object has accepts() virtual function for this purpose. But I wasn't able to use it to my advantage, because I foresee a need to bypass the check sometimes (to store None, for example), and using accepts() won't let me bypass it.

Reply:
If this is the case, then you should prefer to not use `virtual` at all until it is actually needed.

That being said - what is the issue with overriding `Py::Object::accepts`? I'm not sure I understand.

Reply:
> If this is the case, then you should prefer to not use virtual at all until it is actually needed.

That's what PyCXX did in their `Py::Object::set(...)`. And I hate them for that, because overriding `set` whould have made updating `cppptr` much easier, and I wouldn't have had to screen off all the ways to change the associated PyObject*.

As for overriding accepts.. that's difficult to explain, I feel like I need the typecheck function separate. 
For example, currently, one can't construct PyHandle from nothing. But I have a feeling I may have to need that. Then, it should be able to be set to None. None won't get through typecheck, but I can bypass it in default constructor.
Bypassing set(), which calls accepts(), is impossible, as PyObject* is a private member of Py::Object, and can't be set in any other way.